source/transfer_list.rst

```diff
@@ -146,15 +146,49 @@ The TE header is defined in :numref:`tab_te_header`.
      - The exact size of the data content in bytes, not including inter-TE padding. May be 0.
 
 
-Entry type ranges
------------------
-
-The content of the data section is determined by the tag id. The tag id space contains two ranges:
-
- #. Standard range, and
- #. Non-standard range
-
-The `tag_id` ranges are described in :numref:`tab_tag_id_ranges`.
+Entry type allocation
+---------------------
+
+Tag IDs must be allocated in this specification before use. A new tag ID can be
+allocated by submitting a pull request to this repository that adds a
+description of the respective TE data layout to this specification. Tag IDs do
+not have to be allocated in order. Submitters are encouraged to try to group
+tag IDs together in logical clusters at 256-aligned boundaries (e.g. all tags
+related to a particular chipset or to a particular firmware project could use
+adjacent tag numbers), but there are no predefined ranges and no reservations
+of tag ranges for specific use.
+
+Tags are expected to have a simple layout (representable by a C structure) and
+each tag should only represent data for a single logical concept. Data for
+multiple distinct concepts should be split across different tags, even if
+they're always expected to appear together on the first platform adding the tag
+(to encourage reusability in different situations). The same tag ID may occur
+multiple times in the TL to represent multiple instances of the same kind of
+object. Tag layouts (including the meaning of all fields) are considered stable
+after being added to this specification and may never be changed in a backwards
+incompatible way. If a backwards-incompatible change is desired, a new tag ID
+should be allocated for the new version of the layout instead.
+(Backwards-compatible ways to change existing layouts could be increasing the
+data_length to append new fields behind the structure or using fields in the
+layout that were previously marked as reserved, as long as the TE would still
+be considered valid for older implementations that ignore these new fields. TE
+layouts which have been changed like that must clearly document which fields
+have been added after the first version.)
```

Comment:
Should we prohibit a version number in the TE? It seems like that could be an invitation to do crazy things.

Also, should we suggest that people can allocate a new tag if the old one has outlived its usefulness?

Reply:
I agree that a version field should not be allowed in the TE definition (maybe we can police/enforce that during the TE PR review).
A newer consumer needs to still be able to detect the older layout.
I expect that to be done via the data_size field.
Are there cases where the data_size is ambiguous in identifying the v1 and v2 of a TE layout?

Reply:
Okay, I rewrote this part a bit to more explicitly define how backwards-compatible changes should be made, and include the version number prohibition. (In practice, of course, version numbers may still be included in cases where a separately defined data structure like an FDT or ChromeOS' vboot work buffer is embedded in a TE.)